### DISCUSSION TITLE
A begine's naive question, please help

### DISCUSSION THREAD

**Discussion by onsosb** [mojolicious/mojo#1937] [Q&A]
I have a fairly complex perl code that runs in batch mode, command line.  I need to build a simple web app to run the code and display the output.  The program, in the command line mode  takes as input a text file utf8 encoding, some optional arguments and displays the output or writes it to a file.
example: perl ./my appl.pl  -input= input.txt   > result.txt
The idea is the web version either gets the file name or reads the input text from a text field, passes the info to the perl code and writes out the  out put file and also displays the result in a text field
Any comments, codes, etc is truly appreciated.  I really need lots of hand holding
Thanks
Bid

**Comment by s1037989** [comment]
See the cookbook for [subprocesses](https://docs.mojolicious.org/Mojolicious/Guides/Cookbook#Subprocesses). You'll need to play around with what you need for the inputs. Everything you should need to be successful with this should be covered by the Tutorial Guide.

```
use Mojolicious::Lite -signatures;
use Mojo::IOLoop;
use Mojo::File qw(tempfile);

get '/' => sub ($c) {
  my $input = tempfile->spurt($c->param('input'));
  Mojo::IOLoop->subprocess->run_p(sub {
    qx(myapp.pl -input=$input);
  })->then(sub (@results) {
    $c->render(text => join '', @results);
  })->catch(sub ($err) {
    $c->reply->exception($err);
  });
};

app->start;
```

**Comment by jhthorsen** [comment]
If you want the actual output, you can capture STDOUT using https://metacpan.org/pod/Mojo::IOLoop::ReadWriteFork

```
$fork->run_and_capture_p('myappl.pl', '-input=input.txt')->then(sub {
  my $asset = shift;
  $c->res->headers->content_type('text/plain');
  $c->reply->asset($asset);
});
```
However if the script runs for a long time and you want to guaranty that it completes you might want to look into https://docs.mojolicious.org/Minion instead.

**Reply by s1037989** [reply]
What are the differences between Subprocess and ReadWriteFork?

**Reply by jhthorsen** [reply]
Subprocess is for forking perl and run a *callback* in that fork (which could contain system()). The return value from the Subprocess sub, will be sent to the parent process.
ReadWriteFork is for running any *program* and piping STDIN/STDOUT/STDERR from and to that process. (ReadWriteFork can also run a *callback*, but it's all about I/O)
